xonsh/environ.py

```diff
@@ -106,7 +106,6 @@ def to_debug(x):
     re.compile('\w*PATH$'): (is_env_path, str_to_env_path, env_path_to_str),
     'PATHEXT': (is_env_path, str_to_env_path, env_path_to_str),
     'RAISE_SUBPROC_ERROR': (is_bool, to_bool, bool_to_str),
-    'RIGHT_PROMPT': (is_string, ensure_string, ensure_string),
```

Comment:
I implemented string_or_callable and applied it to $PROMPT and $MULTILINE_PROMPT, please take a look.

I considered adding another validator to check the return type of a callable, but I'm not sure there is much value in that since the function the user assigns could, depending on global state[1], return a string or something else. I can add that validator if you think there is value in that after all.

[1] My particular function is this:
```
def _retcode():
    if __xonsh_history__.rtns:
        ret_code = __xonsh_history__.rtns[-1]
        if ret_code:
            return str(ret_code)
    return ''

$FORMATTER_DICT['retcode'] = _retcode

def right_prompt():
    retcode = _retcode()
    if retcode:
        return '{YELLOW}{BACKGROUND_RED}{retcode}{NO_COLOR}'
    else:
        return '{branch_color}{curr_branch}{NO_COLOR}'
```
so I think depending on global state will be a common use case.